pkg/odo/cli/component/watch.go

```diff
@@ -292,3 +298,38 @@ func NewCmdWatch(name, fullName string) *cobra.Command {
 
 	return watchCmd
 }
+
+// regenerateAdapterAndPush is used as a DevfileWatchHandler in WatchParameters; it is a wrapper around adapter.Push()
+// that first regenerates the component adapter before calling push. This ensures that it has picked up the latest
+// devfile.yaml changes
+func regenerateAdapterAndPush(pushParams common.PushParameters, watchParams watch.WatchParameters) error {
+	var adapter common.ComponentAdapter
+	adapter, err := regenerateComponentAdapterFromWatchParams(watchParams)
+	if err == nil {
+		err = adapter.Push(pushParams)
+	}
+
+	return err
+}
+
+// regenerateComponentAdapterFromWatchParams (re)generates a component adapter from the given watch parameters.
+func regenerateComponentAdapterFromWatchParams(parameters watch.WatchParameters) (common.ComponentAdapter, error) {
+
+	// Parse devfile and validate
+	devObj, err := devfile.ParseAndValidate(parameters.DevfilePath)
+	if err != nil {
+		return nil, err
```

Comment:
Should we wrap this error with a message?

Reply:
Done.